diff -durN libxml2-2.5.7.orig/xmlreader.c libxml2-2.5.7/xmlreader.c
--- libxml2-2.5.7.orig/xmlreader.c	Thu Apr 24 13:12:32 2003
+++ libxml2-2.5.7/xmlreader.c	Fri May 23 14:26:08 2003
@@ -688,6 +688,8 @@
     do {
         if (xmlTextReaderGetSuccessor(reader->node) != NULL)
 	    return(1);
+        if (reader->ctxt->nodeNr <= reader->depth)
+	    return(1);
 	if (reader->mode == XML_TEXTREADER_MODE_EOF)
 	    return(1);
 	val = xmlTextReaderPushData(reader);
